src/common/transformations/src/transformations/common_optimizations/pull_through_reduce.cpp

```diff
@@ -0,0 +1,235 @@
+        std::iota(std::begin(result) + begin_pos, std::end(result), end_pos);
+        return result;
+    }
+    return {};
+}
+
+// Update given reshape_input_shape by inserting "1" dimension on the postion represented by axes_to_insert
+std::shared_ptr<ov::opset9::Constant> update_reshape_target_shape(const ov::Shape& reshape_input_shape,
+                                                                  const std::vector<int64_t>& axes_to_insert) {
+    auto result = std::vector<int64_t>(std::begin(reshape_input_shape), std::end(reshape_input_shape));
+    for (const auto& axis : axes_to_insert) {
+        result.insert(std::next(std::begin(result), axis), 1);
+    }
+    return ov::opset9::Constant::create(ov::element::i64, ov::Shape{result.size()}, result);
+}
+
+// Return true if given inputs have some common elements, otherwise return false.
+bool have_same_axes(const std::vector<int64_t>& unsqueeze_axes, const std::vector<int64_t>& reduce_op_axes) {
+    return std::find_first_of(std::begin(unsqueeze_axes),
+                              std::end(unsqueeze_axes),
+                              std::begin(reduce_op_axes),
+                              std::end(reduce_op_axes)) != std::end(unsqueeze_axes);
+}
+}  // namespace
+
+ov::pass::PullUnsqueezeThroughReduce::PullUnsqueezeThroughReduce() {
+    MATCHER_SCOPE(PullUnsqueezeThroughReduce);
+
+    const auto input = pattern::any_input(pattern::has_static_rank());
+    const auto unsqueeze_axes = pattern::wrap_type<opset9::Constant>();
+    const auto unsqueeze = pattern::wrap_type<opset9::Unsqueeze>({input, unsqueeze_axes}, pattern::has_static_rank());
+    const auto reduce_axes = pattern::wrap_type<opset9::Constant>();
+    const auto arithmetic_reduce = pattern::wrap_type<op::util::ArithmeticReductionKeepDims>({unsqueeze, reduce_axes});
+    const auto logical_reduce = pattern::wrap_type<op::util::LogicalReductionKeepDims>({unsqueeze, reduce_axes});
+    const auto reduce = std::make_shared<pattern::op::Or>(OutputVector{arithmetic_reduce, logical_reduce});
+
+    matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        auto& pattern_map = m.get_pattern_value_map();
+        const auto input_node = pattern_map.at(input);
+        const auto reduce_node = pattern_map.find(arithmetic_reduce) != std::end(pattern_map)
+                                     ? pattern_map.at(arithmetic_reduce).get_node_shared_ptr()
+                                     : pattern_map.at(logical_reduce).get_node_shared_ptr();
+        const auto unsqueeze_node = pattern_map.at(unsqueeze).get_node_shared_ptr();
+        auto unsqueeze_axes_input =
+            std::dynamic_pointer_cast<opset9::Constant>(pattern_map.at(unsqueeze_axes).get_node_shared_ptr());
+        auto reduce_axes_input =
+            std::dynamic_pointer_cast<opset9::Constant>(pattern_map.at(reduce_axes).get_node_shared_ptr());
+
+        auto unsqueeze_axes_val = unsqueeze_axes_input->cast_vector<int64_t>();
```

Comment:
you didn't check for nullptr before using shared pointer. code analyzer would be mad at this :)

Reply:
already discussed with @praasz (https://github.com/openvinotoolkit/openvino/pull/13047#discussion_r995435984) I'm adding checks again to move the PR forward, but let's discuss offline how to test it locally or maybe change the analyzer policy